src/Compilers/Core/Portable/Syntax/ChildSyntaxList.Enumerator.cs

```diff
@@ -41,7 +48,7 @@ internal void InitializeFrom(SyntaxNode node)
             public bool MoveNext()
             {
                 var newIndex = _childIndex + 1;
-                if (newIndex < _count)
+                if (unchecked((uint)newIndex < (uint)_count))
```

Comment:
i don't understand all these unchecked uint comparisons.  Can you either revert, or add data showing why this is beneficial.

Reply:
This change was made to allow scenarios like the one where a node is not found in a parent
and `-1 & int.MaxValue` is passed in as the initial index of a segment and enumerator.
As such, passing in an arbitrarily large integer here like `-1 & int.MaxValue` (which equals `int.MaxValue`)
to represent an empty segment is allowed despite the increment of it being negative (`int.MaxValue + 1 < 0`)

The comparison casts both to unsigned integers b/c we want to stop if `_childIndex` was less than -1 to begin with,
and it's unchecked to ensure the operation isn't checked for integer overflow (a potential overflow is the intent).
It's in a perf-critical section, it's a single comparison instead of two, and it's a [common pattern.](https://github.com/search?q=repo%3Adotnet%2Froslyn+OR+repo%3Adotnet%2Fruntime+%2F%5CWunchecked%5C%28%5Cs*%5C%28uint%5C%29.*%3F%5B%5Cs%5Cw%5D%28%3C%7C%3E%29%5B%5Cs%5Cw%5D%2F&type=code)

Reply:
whatever the reason, this needs:

1. to be heavily doc'ed.
2. to have its scenario actually tested to ensure it is doing the right thing.  

> we want to stop if _childIndex was less than -1 to begin with,

Can this happen?  Do you have a test which ends up exercising this.

Reply:
> It's in a perf-critical section